src/Microsoft.DotNet.Darc/src/Darc/Operations/GatherDropOperation.cs

```diff
@@ -0,0 +1,576 @@
+                Build assetBuild = null;
+
+                // Figure out whether we've seen this dependency before by looking it up in the dependency
+                // detail map.
+                if (dependencyCache.TryGetValue(dependency, out Build existingBuild))
+                {
+                    assetBuild = existingBuild;
+                }
+                else
+                {
+                    // Go to the BAR to find out where the asset came from.  Every time we look up a new build,
+                    // cache its assets in the map.
+
+                    // Look up the asset by name and version.
+                    Console.WriteLine($"Looking up {dependency.Name}@{dependency.Version} in Build Asset Registry...");
+                    IList<Asset> matchingAssets = await rootBuildRemote.GetAssetsAsync(dependency.Name, dependency.Version);
+
+                    // Because the same asset could be produced more than once by different builds (e.g. if you had
+                    // a stable asset version), look up the builds by ID until we find one that built the right commit.
+                    foreach (var asset in matchingAssets)
+                    {
+                        Console.WriteLine($"Looking up build {asset.BuildId.Value} in Build Asset Registry...");
+                        Build potentialBuild = await rootBuildRemote.GetBuildAsync(asset.BuildId.Value);
+                        // Do some quick caching after this lookup.
+                        foreach (Asset buildAsset in potentialBuild.Assets)
+                        {
+                            dependencyCache.Add(
+                                new DependencyDetail() {
+                                    Name = buildAsset.Name,
+                                    Version = buildAsset.Version,
+                                    Commit = potentialBuild.Commit,
+                                    RepoUri = potentialBuild.Repository },
+                                potentialBuild);
+                        }
+
+                        // Determine whether this build matches. Commit should be enough.  We could
+                        // also test for repo uri here but I don't think its necessary.
+                        if (potentialBuild.Commit == dependency.Commit)
+                        {
+                            assetBuild = potentialBuild;
+                            break;
+                        }
+                    }
+                }
+
+                    
+                if (assetBuild == null)
+                {
+                    errors.Add($"Could not find build that produced {dependency.Name}@{dependency.Version} @ {dependency.Commit}");
+                    if (!_options.ContinueOnError)
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    if (!builds.Contains(assetBuild))
```

Comment:
>if (!builds.Contains(assetBuild)) [](start = 20, length = 33)

Since `builds` is a HashSet you don't need to check if it contains a value. If it does it will just move on without throwing